### DISCUSSION TITLE
Debug MCU via node-red editor

### DISCUSSION THREAD

**Discussion by ralphwetzel** [phoddie/node-red-mcu#13] [Ideas]
@phoddie 
When running a flow, the NR runtime sends messages back to the editor to display display debug data.
With reference to xsbug, I see - for this feature - a significant overlap. It seems to me, that the MCU (Simulator) send messages back to xsbug, that are displayed there.
If there was a documentation of the protocol, I could create a logic that connects to the MCU (Simulator) & catches this information - to display it in the editor. This could eliminate the need to operate with an additional tool & make life easier for people...
Of course xsbug provides additional information & options - which yet might only be necessary for experienced operators.

**Comment by phoddie** [comment]
@ralphwetzel – Having a debugger for JavaScript is a feature. I personally cannot imagine developing a project of any complexity without one. We have that feature. I certainly wouldn't want to take it away.

I understand that Node-RED generally runs without a JavaScript debugger so people working with Node-RED are accustomed to that. To maintain that experience, an option to run without xsbug would be fine.

The Node-RED editor provides feedback about running projects. As I understand it, that is done using the `status` set on nodes. The Node-RED MCU runtime has that information. It could relay that back to the editor. That's not trivial. The communication path used by xsbug varies by device and simulator. It can be serial (ESP32, for example) or a TCP socket (simulator, for example). 

I don't know that JavaScript debugging and display of status in the Node-RED editor need to be mutually exclusive. It seems possible that they could co-exist and that it would be valuable if they did.

Would you consider inverting your suggestion? Rather than starting with the xsbug protocol, let's start by understanding the data that get sent to the Node-RED Editor for status and how that is communicated to Node-RED. Since we are not running inside the Node runtime, that's going to be quite different.

**Comment by ralphwetzel** [comment]
@phoddie
You intension to begin with implementing the functionality of the .status() feature looks like a good idea. I briefly checked the NR source - and both topics, .status() & debug, seem to be implemented quite similar. Could you please provide a small flow that we use as test bed? I anticipate that there will be work to be done on both ends of the interface...

**Comment by ralphwetzel** [comment]
Ok. Here's the proposal of how to implement the .status() functionality:

**MCU end:**
The [Node status](https://nodered.org/docs/creating-nodes/status) is an object of three properties (`fill`, `shape`, `text`). In addition, I need the `id` of the node that is sending it's status; this info is present in the `flows.js` JSON:
```
{
  "id": "e101f979cfca0e99",
  "status": {
    "fill": "green",
    "shape": "dot",
    "text": "I feel good..."
  }
}
```

This object shall be sent (e.g. JSONized) to a server running in NR that is awaiting incoming calls. I propose to send this initially via websockets - if feasable on MCU side. Host & port data could be (transparently) provided to MCU as part of the build process. When this is operational w/ the sim, we need to find a way of how to integrate a real hardware. Do you have a proposal for a setup? 

**NR end:**
As part of the build process, the mentioned server will be opened. Incoming status data will then be injected into the NR system.
I would prefer to be able to kill the instances of `mcsim` & `xsbug` spawned as part of the build process - to ensure there's always a clean test environment. Do you see an option to get their PID's returned somehow?

**Comment by phoddie** [comment]
I wouldn't recommend using the Wi-Fi connection to carry the status information. There are some problems:

1. Not all MCUs have a network connection. The Raspberry Pi Pico and nrf52 are both examples of projects that could run Node-RED MCU but don't have Wi-Fi.
2. Buffering. When sending data over Wi-Fi, eventually the internal send buffers of the MCU fill and no more data can be sent until they empty. Managing that is complex and potentially uses quite a bit of memory (which is limited).

Carrying the status (and debug) data over the xsbug connection avoids both of these problems because the debugging communication with an MCU usually runs over serial.

Once the status / debug data is available over the xsbug connection, then it can be relayed to Node-RED as you describe.

**Comment by ralphwetzel** [comment]
It's a very fair point to consider that MCUs (usually) use a serial connection for debugging purposes. In that case, we should concentrate on a serial protocol rather than demanding a "man in the middle" setup. If an operator yet likes to use `xsbug`, relaying status & (later) debug back to NR should keep a consistent experience.

**Comment by phoddie** [comment]
> In that case, we should concentrate on a serial protocol...

But the simulator connects to xsbug using a TCP connection. ;) A general solution needs to support both. The xsbug connection is the one place the Moddable SDK guarantees that a connection is available. From a compatibility and maintenance perspective, it is the most direct path. It also ensures that the JavaScript debugger is available when desired.

**Comment by phoddie** [comment]
FWIW – I have an idea for how to implement this which I think will work well for both of our goals. I need to think about it a bit more before trying to explain.

**Reply by ralphwetzel** [reply]
👍

**Comment by phoddie** [comment]
@ralphwetzel, I've committed the skeleton of a possible solution.
- Works with the simulator
- Works with all MCUs
- Re-uses existing xsbug communication
- Allows Node-RED to receive messages (status, debug, etc)
- Allows xsbug to operate as usual for debugging JavaScript
- Works without xsbug

The implementation is at `tools/xsbugproxy`. As the name suggests, it is a proxy for the xsbug connection. It is written in Node. The Moddable SDK project connects to the proxy. The proxy optionally connects to xsbug. I've tested the proxy running stand-alone. I assume you would update it so it relays messages to the Node-RED runtime.

The example has comments that attempt to explain the details. It is a little rough, but hopefully you can work through it.

N.B. The proxy does not currently extract any messages (status or otherwise). There is a comment where that should be done. Note that message may be fragmented, especially when arriving from an MCU, so the message parsing will also need to potentially combine fragments to create a complete message.

**Reply by ralphwetzel** [reply]
@phoddie, this sounds like a promising concept.
Let me check how to integrate it...

**Reply by ralphwetzel** [reply]
https://github.com/phoddie/node-red-mcu/blob/a9aebf7e54e7539907b6fb7f97bd78c118a29ef0/tools/xsbugproxy/index.js#L5-L14

There's already a [logic implemented](https://github.com/Moddable-OpenSource/moddable/blob/86c6dff2cad47b2b72ad52e86de9eb8acfbcb8c0/xs/platforms/mac_xs.c#L161-L164
) to alter the port from `5002` to `5003`.
Would it be possible to trigger this change on demand & "reuse" this logic ... eliminating the need to patch the code to support the proxy?

**Reply by phoddie** [reply]
@ralphwetzel - Good find. It is there to allow debugging of xsbug (which is written in JavaScript using the Moddable SDK) with xsbug. The trigger is the name of the executable. My feeling is we need something more general but I'm not sure yet. Either way it should be too difficult.

Also, to use this proxy with an MCU requires changing the port used by `serial2xsbug` similar to the change for macOS. I'll track that down later.

**Reply by ralphwetzel** [reply]
https://github.com/phoddie/node-red-mcu/blob/a9aebf7e54e7539907b6fb7f97bd78c118a29ef0/tools/xsbugproxy/index.js#L65-L68
<strike>Unfortunately the info sent is not sufficient to display the status next to the "emitting" node in the editor. What I need (in the message sent) is the ID that identifies this node within NR. As said, this ID is available in the JSON that is provided as `flows.js`. I propose to hand this over as a standard property when creating the instance of the class / node ... and then have it at hand when necessary.</strike>

Sorry. I was listening to `debug`... `status()` already sends the right data!

**Reply by phoddie** [reply]
I think it would help me to have an example of a simple flow that generates status and the JSON message it generates in Node-RED for status. That would give something to emulate and ensure we are looking at the same thing. Would you provide that?

By the way, were you able to run the proxy successfully?

**Comment by phoddie** [comment]
OK, great! `status` is correct. How about `debug`?

**Reply by ralphwetzel** [reply]
@phoddie:
> By the way, were you able to run the proxy successfully?

Yes. Great idea...

> OK, great! status is correct. How about debug?

This is how `status` looks like:
```
   {
      xsbug: {
        bubble: {
          _name: '',
          _value: '1',
          _path: '/Users/ralph/Projekte/node-red-mcu-test/nodered.js',
          _line: '237',
          __text: '{"status":{"text":1658081729238,"source":{"id":"799b7e8fcf64e1fa","type":"debug","name":"debug 4"}}}',
          toString: [Function: toString]
        }
      }
    }
```

And here's an example for `debug`:
```
    {
      xsbug: {
        bubble: {
          _name: '',
          _value: '2',
          _path: '/Users/ralph/Projekte/node-red-mcu-test/nodered.js',
          _line: '325',
          __text: '{"payload":1658081681239,"topic":"","test":"tester","xxx":true}',
          toString: [Function: toString]
        }
      }
    }
```
As you can see, `status` comes with `source` - which holds the info necessary to get the "emitting" node.

**Reply by phoddie** [reply]
I've update the debug node so it includes a `source` property on the JSON that goes to the sidebar. That should give you what you need.

**Comment by ralphwetzel** [comment]
Proof of concept - that still needs a lot of polishing:

https://user-images.githubusercontent.com/16342003/179423471-5f27d0d1-5eed-4bf0-8be4-d647bf430b33.mov

Would it be possible to send a message through the proxy to the MCU ... to e.g. forward a button press on an inject node to the flow running on the MCU?

**Reply by phoddie** [reply]
Great work! That's a lot of different pieces to have working together.

> Would it be possible to send a message through the proxy to the MCU ... to e.g. forward a button press on an inject node to the flow running on the MCU?

Yes, it should be possible to relay a message over the xsbug connection to the MCU. Once the message gets to the MCU, there is some additional runtime work needed to dispatch it to the node, and implement the expected behavior. The details will require some experimentation.

**Comment by phoddie** [comment]
@ralphwetzel – There's an update that supports triggering an inject node over the xsbug connection.

- The implementation is easily extensible to send other triggers as needed
- xsbugproxy is updated to show how to send the inject trigger
- Works in the simulator. It will work on MCUs too, with a small Moddable SDK update

**Reply by ralphwetzel** [reply]
👍 I'll check how to utilize this...

**Comment by ralphwetzel** [comment]
@phoddie 
It lasted a bit longer than I expected - but here it is: I've created a plugin to integrate this great add-on into the node-red world.
Be aware that it's currently in the state of a PoC & not at all stable.
You'll find some words of what it does & how to install it in the repo: [node-red-mcu-plugin](https://github.com/ralphwetzel/node-red-mcu-plugin).
The PR I raised some minutes ago is necessary to enable the status output of the debug node (of the node-red-mcu side).

**Reply by phoddie** [reply]
@ralphwetzel – This is very exciting! I will make time to check it out in the next day or two and let you know how it goes. Thank you!

**Reply by phoddie** [reply]
@ralphwetzel – apologies for being a little slow in getting to this. Very soon, I hope.

**Reply by phoddie** [reply]
In getting started, I've bumped into some problems.

- mcu_plugin.js requires "ws". This fails (presumably because the dependency isn't installed). It looks like `ws` is unused, so I just commented out the line.
- The path check at line 46 fails because on my system `require.main.path` results to `/usr/local/lib/node_modules/node-red` which doesn't pass the test. The path constructed at line 47 is incorrect on my system as well. Here's the modified block:
```
    if (rmpd[rmpdl-1]=="node_modules" /* && rmpd[rmpdl-2]=="packages" && rmpd[rmpdl-3]=="node-red" */) {
        let p = path.join(rm,"node_modules","@node-red","runtime","lib", "flows", "util.js");
        flowUtil = require(p)
    }
```

That allows Node-RED to launch without an error. It shows:

```
MCU loaded.
MCU added.
```

When triggering the "MCU Builder" node, it generates an error that I'm not sure how to resolve.

```
30 Jul 14:47:09 - [error] [node-red-mcu:d591bc9819311b21] TypeError: Cannot read properties of undefined (reading 'flows')
```

Can you help?

**Reply by ralphwetzel** [reply]
> Can you help?

I'll try to...

> When triggering the "MCU Builder" node, it generates an error that I'm not sure how to resolve.

There's no need to utilize the `MCU Builder` node (currently): You'll find an additional panel in the right hand side panel, labeled MCU.

<img width="486" alt="image" src="https://user-images.githubusercontent.com/16342003/182023589-6b608d85-3351-42c4-b194-10fefc0d1350.png">

This panel allows to select the flows to be built (you need to deploy after the selection!) & has a button to run the building process.

Currently I'm testing with a very simple flow - which is as well provided in the repo:

<img width="455" alt="image" src="https://user-images.githubusercontent.com/16342003/182024102-de1ac2d9-2f5f-4f72-86b5-a9b90f1457d6.png">

**Reply by phoddie** [reply]
Thank you! I completely overlooked the very clear sentence in your documentation, "The plugin adds a side panel...". The side-panel does launch a build with the simulator. I made some breaking changes (uncommitted) that prevent me from seeing it work. I'll resolve that.

Any ideas on what to do about the path differences on our systems that is causing errors at Node-MCU launch?

**Reply by ralphwetzel** [reply]
> Any ideas on what to do about the path differences on our systems that is causing errors at Node-MCU launch?

I already tried to fix that with a commit earlier today...

**Reply by phoddie** [reply]
It is really working! I was able to see Status changes and also use Trigger. I have some non-trivial flows that I use to test Node-RED MCU and they work well with the proxy.

Your proof-of-concept is very promising. In reading the code, I see that it matches the $MODDABLE sources directly. I think we should add a way to set the xsbug port from the command line when building so that this dynamic source code patch is unnecessary. I can look into that.

I think you are also patching Node-RED? I'm not sure how to eliminate that.

In reading the proxy code, I see a couple of possible improvements (these are most likely to matter when communicating with an MCU because the TCP data is more likely to be fragmented):

- `this.inCache += data.toString().trim();` The `trim` could cause a problem if it removes semantically meaningful white space. I suggest omitting it.
- the extraction of the XML fragment could be incorrect in some cases. I think something like this should be more reliable:
```js
				const start = this.inCache.indexOf("<xsbug>");
				const end = (start < 0) ? -1 : this.inCache.indexOf("</xsbug>", start); 
				if (end > 0) {
					const xml = this.inCache.slice(start, end + 8);
					let doc = this.x2js.xml2js(xml);
					this.inCache = this.inCache.slice(end + 8);
```

A small cosmetic detail: the button to run the build is almost hidden on my computer. That confused me a bug. Here a picture so you can see.

![button](https://user-images.githubusercontent.com/6856458/182284471-63eb8cfa-be4d-40a9-bcbf-64e31644de23.png)

**Reply by ralphwetzel** [reply]
> It is really working!

😄 

> I think we should add a way to set the xsbug port from the command line when building so that this dynamic source code patch is unnecessary. I can look into that.

👍 That would make things much easier.

> I think you are also patching Node-RED? I'm not sure how to eliminate that.

No need to eliminate that... for the time being.

> * the extraction of the XML fragment could be incorrect in some cases. I think something like this should be more reliable:

👍 Changed that. Thank you for the proposal.

> the button to run the build is almost hidden on my computer.

I'm currently working on improving the UI...

**Reply by phoddie** [reply]
@ralphwetzel – I see you applied the XML change already. That was quick.

I'm working on the xsbug port support and look forward to the UI improvements.

**Reply by phoddie** [reply]
@ralphwetzel – today's update to the Moddable SDK implements a new `-x` option to `mcconfig` to set the xsbug host and port. This should eliminate the need to patch the Moddable SDK sources to use the xsbug proxy. Here's an example:

```
mcconfig -d -m -p sim -x localhost:5004
```

Be sure to [rebuild tools after updating](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#mac-update). This new capability is implemented for the following:

- macOS simulator
- Linux simulator
- ESP8266 builds on macOS and Linux
- ESP32 builds on macOS and Linux
- Raspberry Pico builds on macOS and Linux

Some notes:

- Windows support is not yet ready. We'll get there.
- The simulator reads the xsbug configuration at launch. If those change, it needs to be relaunched to use the new configuration.
- If `-x` is not present, xsbug defaults to "localhost:5002" as before.

**Reply by ralphwetzel** [reply]
> to set the xsbug host and port

This is great news!

> * The simulator reads the xsbug configuration at launch. If those change, it needs to be relaunched to use the new configuration.

Any reason why this is necessary for the sim? The proxy operation should have no impact on the sim - to my understanding... ?

**Reply by phoddie** [reply]
> Any reason why this is necessary for the sim? The proxy operation should have no impact on the sim - to my understanding... ?

@ralphwetzel – I'm not certain I understand. Node-RED MCU project run on your computer (macOS, Windows, Linux) use the simulator (mcsim process). When use your MCU Build tool inside the Node-RED editor, it launches the mcsim to run the flows and that connects to the proxy.

**Reply by ralphwetzel** [reply]
> I'm not certain I understand.

Never mind. I just didn't get why the sim needs to know anything about xsbug...

**Reply by phoddie** [reply]
No problem. If there's something that still doesn't make sense, let me know. I'm happy to try to explain.

**Reply by ralphwetzel** [reply]
> Be sure to [rebuild tools after updating](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/Moddable%20SDK%20-%20Getting%20Started.md#mac-update). This new capability is implemented for the following:
> 
> * macOS simulator
> * Linux simulator
> * ESP8266 builds on macOS and Linux
> * ESP32 builds on macOS and Linux
> * Raspberry Pico builds on macOS and Linux
> 
> Some notes:
> 
> * Windows support is not yet ready. We'll get there.

@phoddie -- Did you find the time to implement this for Windows as well? I managed to build a flow there successfully ... but it's lacking the features to interact with the editor. I already spent some time checking for errors ... and then remembered this exchange from the early days of our journey! 😄

**Reply by phoddie** [reply]
No progress here, but the notes above aren't quite correct either. Some details:

- For embedded build target (ESP8266, ESP32, Raspberry Pi Pico) Node-RED MCU Edition supports everything that the Moddable SDK supports:
   - ESP8266 - macOS, Windows, Linux
   - ESP32 - macOS, Windows, Linux
   - Raspberry Pi Pico - macOS, Linux
- The Moddable SDK simulator (mcsim) works on macOS, Windows, and Linux. However, the ECMA-419 network classes (TCP, Listener, and UDP) are not yet implemented on Windows. That's very do-able but it is some work.

For any further discussion about this, let's open a new topic, since this one is pretty far from where it started.

**Comment by ralphwetzel** [comment]
@phoddie Is there a way to enumerate (all) the possible platform identifiers (like `esp32/moddable_two`)?

**Reply by phoddie** [reply]
Yes, there is. It is imperfect, because the details are actually left to each port and how they define their manifest. However, by convention, this works:

- The device specific bits for each port are stored in `$MODDABLE/build/devices`. You can enumerate the directories there to get a list of platforms (e.g. `esp`, `esp32`, etc).
- Within each directory there is a `targets` directory (e.g. `$MODDABLE/build/devices/esp32/targets`) with a directory for each subplatform (`moddable_two`, `m5stack_fire`, etc).
- You can assemble those two parts to get platform identifiers (e.g. `esp32/moddable_two`)

Note that the developer may not have their local environment set-up for a given platform. Only those platforms which have tools installed will work.

FWIW – this is something that I recall @hipsterbrown bumped into with [xs-dev](https://github.com/HipsterBrown/xs-dev) too. Any "real" solution should take your work and xs-dev into consideration.

**Reply by ralphwetzel** [reply]
Thank you for that hint. I hoped for a built-in function ... yet this is do-able as well.

> Only those platforms which have tools installed will work.

Is there a way to verify this - checking e.g. the presence of a file/folder?

**Reply by phoddie** [reply]
> Is there a way to verify this - checking e.g. the presence of a file/folder?

Unfortunately, no. The tool set-up for the embedded targets is complex, and each one is different (silicon manufacturer specific!) There's no easy way to verify that they are set-up correctly. `xs_dev` might be able to provide an indication when it is used for set-up of the embedded tools.

**Reply by phoddie** [reply]
>  I hoped for a built-in function ... 

The best I think we could do is to automate what I described in a Moddable SDK tool. That would be more convenient for you and xs-dev. In the future, that could be updated if other mechanisms become available.

**Reply by ralphwetzel** [reply]
👍

**Reply by HipsterBrown** [reply]
In xs-dev, I opted for the quickest solution of manually creating the list of available platforms: https://github.com/HipsterBrown/xs-dev/blob/main/src/commands/run.ts#L33-L70

There is no environment validation when providing those options within xs-dev at the moment; however, this should not be too difficult to implement given some knowledge of the tooling required for various targets, i.e. making sure the Espressif SDK is available for ESP8266. 

Dynamically creating the list of available device / simulator targets could also be accomplished in a similar manner as the list of Moddable examples: https://github.com/HipsterBrown/xs-dev/blob/main/src/commands/run.ts#L89-L97 🤔

**Reply by phoddie** [reply]
> I opted for the quickest solution of manually creating the list of available platforms

That's very pragmatic. The list doesn't change very often, so that's manageable.

> There is no environment validation when providing those options within xs-dev at the moment; however, this should not be too difficult to implement given some knowledge of the tooling required for various targets, i.e. making sure the Espressif SDK is available for ESP8266.

This will always be a bit fragile but as a first-order filter it should work pretty well.

**Comment by ralphwetzel** [comment]
@phoddie 
> I'm working on the xsbug port support and look forward to the UI improvements.

The UI update is [available](https://github.com/ralphwetzel/node-red-mcu-plugin) now. I'm not sure how this plays with your latest changes though; going to check this as a next step...

**Reply by phoddie** [reply]
Your UI update sure looks comprehensive. I can't wait to try it out.

My latest changes should only require a small update on your part – a simplification to how you are outputting the JSON data. Let me know if you have any questions.

**Reply by ralphwetzel** [reply]
After updating & rebuilding the `Moddable SDK` and integrating the latest `node-red-mcu` repository, I get the following error **when the sim launches**, after performing the build process e.g. via `mcconfig -d -x localhost:5004 -m -p sim/m5paper`:

`dlopen(/Users/ralph/Library/Application Support/tech.moddable.mcsim/mc.so, 0x0006): symbol not found in flat namespace '_xs_nodered_util_generateId'`.

Any hints?



> small update on your part – a simplification to how you are outputting the JSON data

I went through the updated documentation, yet couldn't detect any change necessary. It still seems to be "fill `flows.json` & run the build". Do I miss anything?

**Reply by phoddie** [reply]
> dlopen(/Users/ralph/Library/Application Support/tech.moddable.mcsim/mc.so, 0x0006): symbol not found in flat namespace '_xs_nodered_util_generateId'.

Maybe you didn't copy [nodered.c](https://github.com/phoddie/node-red-mcu/blob/main/nodered.c)?

> It still seems to be "fill flows.json & run the build". Do I miss anything?

Two letters: that used to be "fill flows.js" and now it is "fill flows.json".

**Reply by ralphwetzel** [reply]
2x 👍. Already committed these changes.

**Reply by phoddie** [reply]
Your latest works well! It has become a very big project (maybe more total lines of code than the core of Node-RED MCU itself?!?!)

The configuration editor is quite an accomplishment. It also reminds me that it is getting to be time to build some nodes that draw to the screen.

A few comments:

- It appears ot be required to Deploy in Node-RED before building for the MCU. If not, a stale flow is used. It took a little while to figure that out. Is it possible to  build the current editor state instead of the most recent deploy?
- The Instrumentation pop-up is a little confusing with two checkboxes to control three options. Maybe it can just be a pop-up with "debug", "release", and "instrumented" with the label "build type"
- When there are several flows, the last flow is not shown in the "Flows to build for MCU" list

Having the MCU build available so easily let me quickly run a bunch of different flows in different environments. That uncovered a few bugs in my code, so I committed a few fixes. ;)

**Reply by ralphwetzel** [reply]
> Is it possible to  build the current editor state instead of the most recent deploy?

No, that's not possible. We need to inform the runtime that MCU is in control for the selected flows - and do some modifications behind the scene. That's why all nodes (in the selected flows) get flagged as changed (blue ball) & the deploy button illuminates as well. Yet this should only be necessary once...

> * When there are several flows, the last flow is not shown in the "Flows to build for MCU" list

The panel is resizable. Could you please confirm, if it's really missing - or just hidden as overflown?

**Reply by phoddie** [reply]
> The panel is resizable. Could you please confirm, if it's really missing - or just hidden as overflown?

You are right. It is there if I resize the panel enough. There's some bug with the scrolling.

> We need to inform the runtime that MCU is in control for the selected flows - and do some modifications behind the scene. That's why all nodes (in the selected flows) get flagged as changed (blue ball) & the deploy button illuminates as well. Yet this should only be necessary once...

That's no problem for now. Maybe it should be shown in the UI somehow, to help people? Eventually we should get some support from the Node-RED Editor to make that smoother.

**Reply by ralphwetzel** [reply]
> There's some bug with the scrolling.

Confirmed. I'll check...

> Maybe it should be shown in the UI somehow, to help people?

Well, it is - as described. In addition to that: The `Build` button should only be active (and trigger-able) if there's at least one flow selected...

**Reply by ralphwetzel** [reply]
> There's some bug with the scrolling.

Fixed now.
